sound/soc/sof/topology.c

```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
-	memset(&config, 0, sizeof(config));
-
-	/* get any common DAI tokens */
-	ret = sof_parse_tokens(scomp, &config, dai_link_tokens,
-			       ARRAY_SIZE(dai_link_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret != 0) {
-		dev_err(scomp->dev, "error: parse link tokens failed %d\n",
-			le32_to_cpu(private->size));
-		return ret;
-	}
-
 	/*
 	 * DAI links are expected to have at least 1 hw_config.
 	 * But some older topologies might have no hw_config for HDA dai links.
 	 */
-	num_hw_configs = le32_to_cpu(cfg->num_hw_configs);
-	if (!num_hw_configs) {
-		if (config.type != SOF_DAI_INTEL_HDA) {
+	hw_config = cfg->hw_config;
+	number_configs = le32_to_cpu(cfg->num_hw_configs);
+	if (!number_configs) {
+		if (config->type != SOF_DAI_INTEL_HDA) {
 			dev_err(scomp->dev, "error: unexpected DAI config count %d!\n",
 				le32_to_cpu(cfg->num_hw_configs));
 			return -EINVAL;
 		}
+		number_configs = 1;
 	} else {
 		dev_dbg(scomp->dev, "tplg: %d hw_configs found, default id: %d!\n",
 			cfg->num_hw_configs, le32_to_cpu(cfg->default_hw_config_id));
 
-		for (i = 0; i < num_hw_configs; i++) {
-			if (cfg->hw_config[i].id == cfg->default_hw_config_id)
+		for (current_config = 0; i < number_configs; current_config++) {
+			if (hw_config[current_config].id == cfg->default_hw_config_id)
 				break;
 		}
 
-		if (i == num_hw_configs) {
+		if (current_config == number_configs) {
 			dev_err(scomp->dev, "error: default hw_config id: %d not found!\n",
 				le32_to_cpu(cfg->default_hw_config_id));
 			return -EINVAL;
 		}
 	}
 
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* reserve memory for all hw configs, eventually freed by widget */
+	config = kzalloc(number_configs * sizeof(*config), GFP_KERNEL);
+	if (!config)
+		return -ENOMEM;
 
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.format = le32_to_cpu(hw_config->fmt);
+	/* get any common DAI tokens */
+	ret = sof_parse_tokens(scomp, config, dai_link_tokens,
+			       ARRAY_SIZE(dai_link_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: parse link tokens failed %d\n",
+			le32_to_cpu(private->size));
+		goto err;
+	}
+
+	config->hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+	config->format = le32_to_cpu(hw_config->fmt);
+
+	/* copy common data to all config ipc structs */
+	for (i = 1; i < number_configs; i++) {
+		config[i].hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+		config[i].format = hw_config[i].fmt;
+		config[i].type = config[i - 1].type;
+		config[i].dai_index = config[i - 1].dai_index;
+	}
 
 	/* now load DAI specific data and send IPC - type comes from token */
-	switch (config.type) {
+	switch (config->type) {
 	case SOF_DAI_INTEL_SSP:
 		ret = sof_link_ssp_load(scomp, index, link, cfg, hw_config,
-					&config);
+					config);
 		break;
 	case SOF_DAI_INTEL_DMIC:
 		ret = sof_link_dmic_load(scomp, index, link, cfg, hw_config,
```

Comment:
this seems to be changing the `hw_config` parameter for all `sof_link_*_load()` functions. You deal with it for SSP but you don't touch others. Before this patch `hw_config` was pointing to the matching HW config:
```
hw_config = &cfg->hw_config[i];
```
now it's the first of them:
```
hw_config = cfg->hw_config;
```
Is this correct? Didn't you mean to change these calls to e.g.
```
ret = sof_link_dmic_load(scomp, index, link, cfg, hw_config + current_config,
```
?

Reply:
I need to parse all hw_configs to be able later to choose 1 of them dynamically. Before none of our topologies included multiple configs. So I guess you could define multiple configs for other dais than SSP -> then we would just discard the others and choose the "default" -> as we did before. So in essence this was previously always sending the first config.

But As I'm doing this for SSP only currently, I'll change the array subscript back for the other dais. If someone wants to add the support for multiple configs for other dais later, the pointer should point to the first config.

Reply:
should match now